### DISCUSSION TITLE
Radio buttons using check and toggle button groups

### DISCUSSION THREAD

**Discussion by ITwrx** [can-lehmann/owlkettle#59] [Q&A] [answered]
It seems in gtk4 we're supposed to use check buttons or toggle buttons in a group to act as radio buttons. Is this already supported in owlkettle, because i'm not seeing it in the examples, nor the source code?

thanks

**Comment by can-lehmann** [comment, marked as answer]
I added a new `RadioGroup` widget in https://github.com/can-lehmann/owlkettle/commit/a5cb51b4b45fa4b03844790ab3a7dde3389560c9 which allows you to use radio buttons. The docs can be found here: https://github.com/can-lehmann/owlkettle/blob/main/docs/widgets.md#radiogroup
I hope to create a longer example soon.

**Reply by ITwrx** [reply]
thanks a lot!

**Reply by ITwrx** [reply]
Well, i tried to use it in my project, but i haven't figured it out yet.  i have more questions than answers, so i'll have to wait on a fuller example. Maybe the CRUD example could have a radio input? Then it would have the benefit of context in an mini-app. Whatever works. Just a thought.

**Reply by can-lehmann** [reply]
I polished the example I used to test the `RadioGroup` widget and uploaded it here: https://github.com/can-lehmann/owlkettle/blob/main/examples/widgets/radio_group.nim

**Reply by ITwrx** [reply]
That example helped to illustrate integration and used a seq for labels, which is nice and appreciated, but it still doesn't show how to set a `value` for each option and get/pass that. An `html radio group`, for example,  has `values` and the `values` are what i would want to pass to my model. `Labels` are not usually suitable for passing as the `value` as they are optimized for the human user.

**Reply by can-lehmann** [reply]
Ah, I see what you mean. The `RadioGroup` widget itself does not support setting values for the individual entries. Instead it just gives you the index of the selected option. You can however use an `enum`. Here is how the model would look for that (warning: code is untested):

```nim
type UserKind = enum
  UserAdmin = "Admin"
  UserModerator = "Moderator"
  UserDefault = "User"
```

And then you could use a `RadioGroup` to select the `UserKind`

```nim
RadioGroup:
  selected = ord(stateVariableWhereUserKindIsStored)
  
  proc select(index: int) =
    stateVariableWhereUserKindIsStored = UserKind(index)

  for kind in low(UserKind)..high(UserKind):
    Label(text = $kind, xAlign = 0)
```

I might extend the CRUD example to include this when I find the time...

**Reply by ITwrx** [reply]
Yeah, i that seemed to be the case when i looked at the upstream gtk widgets and the owlkettle radio widget source code, but wasn't sure what to do about it.  

Regarding the possible CRUD extension: it wouldn't hurt my feelings to see the whole flow of using a radio group/groups integrated with the other fields in a dialog and the model it's altering.  

Don't feel rushed/obligated though. I can work on another part of the project.

thanks

**Reply by can-lehmann** [reply]
I found some time to quickly add this to the CRUD example: https://github.com/can-lehmann/owlkettle-crud/commit/0328224f9138b73aad347e1b068d321367bed63b

**Reply by ITwrx** [reply]
sweet! i'll take a look soon.

**Reply by can-lehmann** [reply]
Note that you will need to delete the existsing database for this to work, since the schema changed (& I did not bother to include migrations).

**Reply by ITwrx** [reply]
yeah, no need for migrations. thanks for the note.
